beacon-chain/state/interface/interface.go

```diff
@@ -0,0 +1,190 @@
+	"github.com/prysmaticlabs/go-bitfield"
+	pbp2p "github.com/prysmaticlabs/prysm/proto/beacon/p2p/v1"
+)
+
+// ReadOnlyValidator defines a struct which only has read access to validator methods.
+type ReadOnlyValidator interface {
+	EffectiveBalance() uint64
+	ActivationEligibilityEpoch() types.Epoch
+	ActivationEpoch() types.Epoch
+	WithdrawableEpoch() types.Epoch
+	ExitEpoch() types.Epoch
+	PublicKey() [48]byte
+	WithdrawalCredentials() []byte
+	Slashed() bool
+	IsNil() bool
+}
+
+// ReadOnlyValidators defines a struct which only has read access to validators methods.
+type ReadOnlyValidators interface {
+	Validators() []*ethpb.Validator
+	ValidatorAtIndex(idx types.ValidatorIndex) (*ethpb.Validator, error)
+	ValidatorAtIndexReadOnly(idx types.ValidatorIndex) (ReadOnlyValidator, error)
+	ValidatorIndexByPubkey(key [48]byte) (types.ValidatorIndex, bool)
+	PubkeyAtIndex(idx types.ValidatorIndex) [48]byte
+	NumValidators() int
+	ReadFromEveryValidator(f func(idx int, val ReadOnlyValidator) error) error
+}
+
+// ReadOnlyBalances defines a struct which only has read access to balances methods.
+type ReadOnlyBalances interface {
+	Balances() []uint64
+	BalanceAtIndex(idx types.ValidatorIndex) (uint64, error)
+	BalancesLength() int
+}
+
+// ReadOnlyCheckpoint defines a struct which only has read access to checkpoint methods.
+type ReadOnlyCheckpoint interface {
+	PreviousJustifiedCheckpoint() *ethpb.Checkpoint
+	CurrentJustifiedCheckpoint() *ethpb.Checkpoint
+	MatchCurrentJustifiedCheckpoint(c *ethpb.Checkpoint) bool
+	MatchPreviousJustifiedCheckpoint(c *ethpb.Checkpoint) bool
+	FinalizedCheckpoint() *ethpb.Checkpoint
+	FinalizedCheckpointEpoch() types.Epoch
+	JustificationBits() bitfield.Bitvector4
+}
+
+// ReadOnlyBlockRoots defines a struct which only has read access to block roots methods.
+type ReadOnlyBlockRoots interface {
+	BlockRoots() [][]byte
+	BlockRootAtIndex(idx uint64) ([]byte, error)
+}
+
+// ReadOnlyStateRoots defines a struct which only has read access to state roots methods.
+type ReadOnlyStateRoots interface {
+	StateRoots() [][]byte
+	StateRootAtIndex(idx uint64) ([]byte, error)
+}
+
+// ReadOnlyRandaoMixes defines a struct which only has read access to randao mixes methods.
+type ReadOnlyRandaoMixes interface {
+	RandaoMixes() [][]byte
+	RandaoMixAtIndex(idx uint64) ([]byte, error)
+	RandaoMixesLength() int
+}
+
+// ReadOnlyEth1Data defines a struct which only has read access to eth1 data methods.
+type ReadOnlyEth1Data interface {
+	Eth1Data() *ethpb.Eth1Data
+	Eth1DataVotes() []*ethpb.Eth1Data
+	Eth1DepositIndex() uint64
+}
+
+// ReadOnlyAttestations defines a struct which only has read access to attestations methods.
+type ReadOnlyAttestations interface {
+	PreviousEpochAttestations() []*pbp2p.PendingAttestation
+	CurrentEpochAttestations() []*pbp2p.PendingAttestation
+}
+
+// ReadOnlyBeaconState defines a struct which only has read access to beacon state methods.
+type ReadOnlyBeaconState interface {
```

Comment:
We should move the state ifaces to the top of the file to make them easier to find